src/passes/MergeFunctions.cpp

```diff
@@ -0,0 +1,609 @@
+  // `functions`.
+  ConstDiff values;
+  // All uses of the parameter in the primary function.
+  std::vector<Expression**> uses;
+
+  ParamInfo(ConstDiff values, std::vector<Expression**> uses)
+    : values(std::move(values)), uses(uses) {}
+
+  // Returns the type of the parameter value.
+  Type getValueType(Module* module) const {
+    if (const auto literals = std::get_if<Literals>(&values)) {
+      return (*literals)[0].type;
+      return Type::i32;
+    } else if (auto callees = std::get_if<std::vector<Name>>(&values)) {
+      auto* callee = module->getFunction((*callees)[0]);
+      return Type(callee->getSig(), NonNullable);
+    } else {
+      WASM_UNREACHABLE("unexpected const value type");
+    }
+  }
+
+  // Lower the constant value at a given index to an expression
+  Expression*
+  lowerToExpression(Builder& builder, Module* module, size_t index) const {
+    if (const auto literals = std::get_if<Literals>(&values)) {
+      return builder.makeConst((*literals)[index]);
+    } else if (auto callees = std::get_if<std::vector<Name>>(&values)) {
+      auto fnName = (*callees)[index];
+      auto heapType = module->getFunction(fnName)->type;
+      return builder.makeRefFunc(fnName, heapType);
+    } else {
+      WASM_UNREACHABLE("unexpected const value type");
+    }
+  }
+};
+
+// Describes the set of functions which are considered as "equivalent" (i.e.
+// only differing by some constants).
+struct EquivalentClass {
+  // Primary function in the `functions`, which will be the base for the merged
+  // function.
+  Function* primaryFunction;
+  // List of functions belonging to a same class.
+  std::vector<Function*> functions;
+
+  EquivalentClass(Function* primaryFunction, std::vector<Function*> functions)
+    : primaryFunction(primaryFunction), functions(functions) {}
+
+  bool isEligibleToMerge() { return this->functions.size() >= 2; }
+
+  // Merge the functions in this class.
+  void merge(Module* module, const std::vector<ParamInfo>& params);
+
+  bool hasMergeBenefit(Module* module, const std::vector<ParamInfo>& params);
+
+  Function* createShared(Module* module, const std::vector<ParamInfo>& params);
+
+  Function* replaceWithThunk(Builder& builder,
+                             Function* target,
+                             Function* shared,
+                             const std::vector<ParamInfo>& params,
+                             const std::vector<Expression*>& extraArgs);
+
+  bool deriveParams(Module* module,
+                    std::vector<ParamInfo>& params,
+                    bool isIndirectionEnabled);
+};
+
+struct MergeFunctions : public Pass {
+  bool invalidatesDWARF() override { return true; }
+
+  void run(PassRunner* runner, Module* module) override {
+    {
+      // Canonicalize locals indices to make comparison easier.
+      PassRunner runner(module);
+      runner.setIsNested(true);
+      runner.add("reorder-locals");
```

Comment:
I think this is not needed as the pass is scheduled in the global post passes, so it is after this function pass has been run. I guess it might help if the pass is run manually by itself, but I don't think it's worth the extra wasted work in the normal case for that.

Reply:
Makes sense. And I confirmed no size regression after removing reorder-locals in this pass.